### PR TITLE
add: add support for git sources

### LINKED ISSUE #4974: Support for ADD <git ref> in Containerfile
It seems Dockerfile supports adding a git repo files via:

```
ADD [--keep-git-dir=<boolean>] <git ref> <dir>
```

https://docs.docker.com/engine/reference/builder/#adding-a-git-repository-add-git-ref-dir

Do we support that? Can we support that?

### PR DESCRIPTION
**Pull request by danishprakash**
#### What type of PR is this?

> /kind feature

#### What this PR does / why we need it:
adds support for git sources with the ADD directive

#### Which issue(s) this PR fixes:

Fixes #4974

#### Does this PR introduce a user-facing change?

```release-note
adds support for git sources with the ADD directive
```

### CODE REVIEW

**Comment by TomSweeneyRedHat** [comment]
Quick fly by review looks OK to me, but you have test issues and a rebase is needed.

**Comment by danishprakash** [comment]
> Quick fly by review looks OK to me, but you have test issues and a rebase is needed.

You meant packit issues? Because tests seems to be fine afaict

File: add.go [hunk 1]
```diff
@@ -74,9 +76,27 @@ type AddAndCopyOptions struct {
...
 }
 
-// sourceIsRemote returns true if "source" is a remote location.
+// gitURLFragmentSuffix matches fragments to use as Git reference and build
+// context from the Git repository e.g.
+//
+//	github.com/containers/buildah.git
+//	github.com/containers/buildah.git#main
+//	github.com/containers/buildah.git#v1.35.0
+var gitURLFragmentSuffix = regexp.MustCompile(`\.git(?:#.+)?$`)
```
Comment (rhatdan) [inline add.go:85]: Not sure this is necessary, and MustCompile on init variables slows things down on startup.

**Comment by rhatdan** [comment]
LGTM other then remove the MustCompile.

@flouthoc @nalind @giuseppe PTAL

File: add.go (hunk 1 above)
Comment (giuseppe) [reply add.go:85]: can we use  `Delayed()` from the `containers/storage/pkg/regexp` package?

File: add.go [hunk 2]
```diff
@@ -413,12 +446,46 @@ func (b *Builder) Add(destination string, extract bool, options AddAndCopyOption
...
+						ChmodDirs:      chmodDirsFiles,
+						ChownFiles:     chownFiles,
+						ChmodFiles:     chmodDirsFiles,
+						StripSetuidBit: options.StripSetuidBit,
+						StripSetgidBit: options.StripSetgidBit,
+						StripStickyBit: options.StripStickyBit,
+					}
+					getErr = copier.Get(cloneDir, cloneDir, getOptions, []string{"."}, writer)
+					closeErr = writer.Close()
+					if renameTarget != "" && renamedItems > 1 {
```
Comment (eriksjolund) [inline add.go:476]: Line 456:
```
renamedItems := 0
```

Line 476:
```
if renameTarget != "" && renamedItems > 1 {
```

Is `renamedItems == 0` when evaluating the if statment in line 476? (It doesn't look like anything is modifying `renamedItems`)

Comment (danishprakash) [reply add.go:476]: Thanks for catching this, fixed.

File: add.go (hunk 1 above)
Comment (danishprakash) [reply add.go]: Using `Delayed()` now, it seems to help with your concern @rhatdan, wdyt?

**Comment by rhatdan** [comment]
LGTM
@flouthoc @giuseppe @nalind @mheon PTAL

**Comment by TomSweeneyRedHat** [comment]
Changes LGTM

I did a rebase which was required, and now the DCO (signing) test is not passing.   Did you forget to sign a recent commit @danishprakash ?

**Comment by danishprakash** [comment]
I switched git emails recently, that was the issue, fixed now.

**Comment by rhatdan** [comment]
Needs a rebase.

File: add.go [hunk 3]
```diff
@@ -413,12 +446,42 @@ func (b *Builder) Add(destination string, extract bool, options AddAndCopyOption
...
-			}()
+			if sourceIsGit(src) {
+				go func() {
+					var cloneDir string
+					cloneDir, _, getErr = define.TempDirForURL(tmpdir.GetTempDir(), "", src)
+
+					writer := io.WriteCloser(pipeWriter)
+					writer = newTarFilterer(writer, func(hdr *tar.Header) (bool, bool, io.Reader) {
+						return false, false, nil
+					})
```
Comment (nalind) [inline add.go:459]: This doesn't seem to do anything.

**Comment by rhatdan** [comment]
/approve
/lgtm
Thanks @danishprakash

**Pull request merged**